src/MiniProfiler.Shared/Timing.cs

```diff
@@ -77,28 +78,76 @@ public Timing(MiniProfiler profiler, Timing parent, string name, decimal? minSav
         /// Gets or sets All sub-steps that occur within this Timing step. Add new children through <see cref="AddChild"/>
         /// </summary>
         [DataMember(Order = 5)]
-        public List<Timing> Children { get; set; }
+        public List<Timing> Children
+        {
+            get
+            {
+                lock(_lockObject)
+                {
+                    return _children == null ? null : new List<Timing>(_children);
+                }
+            }
+            set
+            {
+                lock(_lockObject)
+                {
+                    _children = value == null ? null : new List<Timing>(value);
+                }
+            }
+        }
 
         /// <summary>
         /// <see cref="CustomTiming"/> lists keyed by their type, e.g. "sql", "memcache", "redis", "http".
         /// </summary>
         [DataMember(Order = 6)]
-        public Dictionary<string, List<CustomTiming>> CustomTimings { get; set; }
+        public Dictionary<string, List<CustomTiming>> CustomTimings
```

Comment:
In my opinion the concurrency issue that i'm trying to fix is highly likely with web applications using async actions. Here is my pr (on top of V4) that reproduces the concurrency issue in a web app: https://github.com/alexsku/dotnet/pull/4

In the code I follow a standard pattern where the action executes multiple async service calls and then awaits for them together using Task.WhenAll. The service calls internally make an async call with ConfigureAwait(false). The completions of those calls would execute in parallel causing the issue.

Also the overhead of the lock is very small when the lock is not contested. I wrote a small test to measure the overhead: https://github.com/alexsku/dotnet/pull/5 and the overhead of lock is negligible (~25 nanoseconds). I think even with the lock the MiniProfiler stays fairly lightweight.